### DISCUSSION TITLE
Antsregistration and mrtransform problem

### DISCUSSION THREAD

**Discussion by mariannamak** [ANTsX/ANTs#1715] [Q&A]
Dear ANTs experts,

I am relatively new to the ANTs community and am currently learning how to use ANTs tools for image registration and transformation. I am encountering some difficulties with antsRegistration and antsApplyTransforms.

My main goal is to register my T2 MR images with the Waxholm atlas (https://scalablebrainatlas.incf.org/rat/PLCJB14) in order to perform volumetric analysis.
The dimensions of my MR images are 384x384x40 with voxel size 0.1x0.1x0.4.
The dimensions of the Waxholm atlas are 512x1024x512 with voxel size 0.04x0.04x0.04.

When I use the rigid + affine + deformable syn (3 stages) approach with antsRegistrationSyN.sh, the resulting registration deforms the neocortex. However, when I switch to sr: rigid + deformable syn (2 stages), the coregistration is relatively successful.
![Screenshot from 2024-03-06 09-51-37](https://github.com/ANTsX/ANTs/assets/166377954/c2359917-ac6e-45fd-9fd4-36dc4a20fcfb)
![Screenshot from 2024-03-06 09-49-53](https://github.com/ANTsX/ANTs/assets/166377954/8642bd1d-2357-4fb1-82fe-dbe1b46556e3)

However, I am encountering issues with AntsApplyTransforms and mrtransform. It seems to be cutting off the whole brain, and I am unsure of the source of this problem.
![Screenshot from 2024-04-08 09-34-14](https://github.com/ANTsX/ANTs/assets/166377954/78fb2748-0971-4e27-82ee-2290cd31e124)

My data is available at the following link: https://drive.google.com/drive/folders/1gh_gPj35v56sp3w8GjDFWTUy6IXT8RJb?usp=drive_link
(MR images, Waxholm atlas, ROIs of wholebrain)

I've tried almost every approach, but I've found that most of my data yields the same results.
Any insights or guidance on resolving these issues would be greatly appreciated.

Thanks in advance,

Best,
Marianna

**Comment by ntustison** [comment]
Please include the precise commands that you used to achieve each of your results.

**Reply by mariannamak** [reply]
1) antsRegistrationSyN.sh -d 3 -m WHS_SD_rat_atlas_v4.nii -f C_V_1_01.nii -o C_V_1_ -i C_V_1_01.txt -n 16 -t sr  
													                                                                         (or s)
2)warpinit ./WHS_SD_rat_atlas_v4.nii C_V_1_identity_warp[].nii
		
3a) antsApplyTransforms -d 3 -i C_V_1_identity_warp0.nii -o ./CV_1_atlas_to_subject0.nii -r ./C_V_1_01.nii -t [./C_V_1_0GenericAffine.mat, 1] -t ./C_V_1_1InverseWarp.nii.gz --default-value 2147483647

3b) antsApplyTransforms -d 3 -i C_V_1_identity_warp1.nii -o ./CV_1_atlas_to_subject1.nii -r ./C_V_1_01.nii -t [./C_V_1_0GenericAffine.mat, 1] -t ./C_V_1_1InverseWarp.nii.gz --default-value 2147483647

3c)antsApplyTransforms -d 3 -i C_V_1_identity_warp2.nii -o ./CV_1_atlas_to_subject2.nii -r ./C_V_1_01.nii -t [./C_V_1_0GenericAffine.mat, 1] -t ./C_V_1_1InverseWarp.nii.gz --default-value 2147483647

4)warpcorrect -quiet -force ./CV_1_atlas_to_subject[].nii  ./CV_1_mrtrix_warp.mif

5)mrtransform  -warp ./CV_1_mrtrix_warp.mif -interp nearest -template ./C_V_1_01.nii ./wholebrain.nii ./CV_1_wholebrain.nii

**Reply by ntustison** [reply]
Okay, a couple issues:

* We don't provide support for external packages, such as mrtrix.  Please limit any questions to ANTs functionality only.  
* The Waxholm atlas has the correct anatomical correction in that if you load it into ITK-SNAP, left/right and anterior/posterior, and inferior/superior axes are all as assumed.  However, as you can see in your screenshots, your data does not conform to this assumption.  The left/right axis lines up but the superior/inferior appears to be swapped with the anterior/posterior axis.  If you don't start with the axes lined up anatomically, your registration attempts won't work.  Perhaps that is what this file ``C_V_1_01.txt`` provides but it's not included and, even if it were, it would be best to perform that prior to any help requests.  
* You should resample any data to a reasonable resolution so we don't have to download and work with extremely large images while trying to debug.
* The Waxholm atlas is skull stripped but your data is not.  We have [mouse brain extraction](https://gist.github.com/ntustison/12a656a5fc2f6f9c4494c88dc09c5621#mouse-brain-extraction) tools in our other packages but, however you do it, this also needs to be matched prior to any registration calls.  

Once you resolve these issues, please post an updated request.

**Reply by mariannamak** [reply]
Hello,
Thank you very much for you time and your advises. Sorry for my late response. I changed the approach of registration, because we still have problems with affine registration. At first, I changed the orientation of my MR scans, but it is still a little bit confusing (we have problems with this orientation in our VnmrJ software, and it's challenging for us)

Now: orientation of atlas: LPI,  orientation of my MR images:  RAI. 
when you open atlas and MR scan in ITK-Snap, the orientations looks same, but it isn´t the same labeling. 
I also moved my MR scans in the space according to the atlas. 

With antsRegistrationSyn, we didn't get the results we were hoping for. However, I tried a different approach using the settings:

antsRegistration --dimensionality 3 --float 0 --collapse-output-transforms 1 --verbose \
  --output [CM4x_ants,CM4x_antsWarped.nii.gz,CM4x_antsInverseWarped.nii.gz] \
  --winsorize-image-intensities [0.01,0.99] --use-histogram-matching 1 \
  --initial-moving-transform [CM4x_RAI_moved.nii, WHS_SD_rat_atlas_v4.nii,1] \
  --transform Rigid[0.1] --metric Mattes[CM4x_RAI_moved.nii, WHS_SD_rat_atlas_v4.nii,1,32,Random,0.5] --convergence [2000x2000x2000x2000x0,1e-6,10] --shrink-factors 16x8x4x2x1 --smoothing-sigmas 4x3x2x1x0vox \
  --transform Affine[0.3] --metric Mattes[CM4x_RAI_moved.nii, WHS_SD_rat_atlas_v4.nii,1] --convergence [10000x10000x10000x10000,1e-6,10] --shrink-factors 8x4x2x1 --smoothing-sigmas 3x2x1x0vox \
  --transform BSplineSyN[0.3,52,0] --metric CC[CM4x_RAI_moved.nii, WHS_SD_rat_atlas_v4.nii,1,4] --convergence [400x200x100x50x25,1e-6,10] --shrink-factors 16x8x4x2x1 --smoothing-sigmas 5x4x2x1x0vox
  
I've experimented with various settings, but these are the ones I've considered the most promising for now.  Some of the registrations appear satisfactory, but at times, I'm not entirely satisfied with the outcomes. It seems to enlarge certain structures in the brain. 
  
  
My questions: 
  1) Is the approach I used correct?

  2) Could I try using different settings in registration, such as adjusting the number of convergence, shrink factor, or smoothing sigmas?

  3) I didn't perform brain extraction. Is it really necessary? 
  
If you have any ideas or tips on how to sort this out, we'd really appreciate it.
Thanks again for your support.

https://drive.google.com/drive/folders/1gh_gPj35v56sp3w8GjDFWTUy6IXT8RJb?usp=drive_link

**Comment by cookpa** [comment]
One issue I see is that you are registering a label image (the Waxholm atlas) to a grayscale image.

I would start by registering the grayscale image to the Waxholm T2* image (available as `WHS_SD_rat_T2star_v1.01.nii.gz`). You can make a mask with

```
ThresholdImage 3 waxholm_mask.nii.gz WHS_SD_rat_atlas_v4.nii.gz 1 Inf
```

The physical space of your moving image is very far off from that of the atlas

<img width="767" alt="image" src="https://github.com/ANTsX/ANTs/assets/611507/b544194a-9e53-4c68-a1fc-2aa23e27af2a">

I used ITK-SNAP's manual registration to roughly align them, then saved the transform. You can use this as an initial transform for antsRegistrationSyN.sh.

```
antsRegistrationSyN.sh -d 3 -f WHS_SD_rat_T2star_v1.01.nii.gz -m C_V_1_01.nii -i initial_transform.txt -x waxholm_mask.nii.gz -t sr -o reg_masked_
```

[initial_transform.txt](https://github.com/ANTsX/ANTs/files/14906910/initial_transform.txt)

**Reply by mariannamak** [reply]
Hello,

Thank you very much for your time and advice. I have successfully resolved the issue with the orientation of my MR images. I tried to create a mask of the MR atlas.

However, I need to clarify that my goal is to align the atlas with my MR images for volumetric evaluation of the MR data. I'm not entirely certain if this approach would be helpful in achieving this goal. While I appreciate any approach suggested, as I am a complete novice in ANTs.
